File: simple-stack-example-mvvm-fragments/src/main/java/com/zhuinden/simplestackexamplemvvm/presentation/paths/tasks/TasksKey.java
package com.zhuinden.simplestackexamplemvvm.presentation.paths.tasks;

import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;

import com.google.auto.value.AutoValue;
import com.zhuinden.simplestackexamplemvvm.R;
import com.zhuinden.simplestackexamplemvvm.application.BaseFragment;
import com.zhuinden.simplestackexamplemvvm.application.BaseKey;
import com.zhuinden.simplestackexamplemvvm.application.injection.Injector;

/**
 * Created by Zhuinden on 2017.07.26..
 */
@AutoValue
public abstract class TasksKey
        extends BaseKey<TasksViewModel> {
    public static TasksKey create() {
        return new AutoValue_TasksKey();
    }

    @Override
    public int navigationViewId() {
        return R.id.list_navigation_menu_item;
    }

    @Override
    public boolean shouldShowUp() {
        return false;
    }

    @Override
    protected void setupFab(Fragment fragment, FloatingActionButton fab) {
        fab.setImageResource(R.drawable.ic_add);
        fab.setOnClickListener(v -> ((TasksFragment) fragment).addNewTask());
    }

    @Override
    protected boolean isFabVisible() {
        return true;
    }

    @Override
    public TasksViewModel newViewModel() {
        return Injector.get().tasksViewModel();
    }

    @Override
    protected BaseFragment createFragment() {
        return new TasksFragment();
    }
}

File: samples/legacy-architecture-samples/simple-stack-example-mvvm-fragments/src/main/java/com/zhuinden/simplestackexamplemvvm/features/tasks/TasksKey.kt
package com.zhuinden.simplestackexamplemvvm.features.tasks


import android.view.View
import androidx.fragment.app.Fragment
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.zhuinden.simplestack.ServiceBinder

import com.zhuinden.simplestackexamplemvvm.R
import com.zhuinden.simplestackexamplemvvm.application.BaseKey
import com.zhuinden.simplestackexamplemvvm.application.injection.Injector
import com.zhuinden.simplestackextensions.servicesktx.add
import kotlinx.android.parcel.Parcelize

/**
 * Created by Zhuinden on 2017.07.26..
 */
@Parcelize
data class TasksKey(private val noArgsPlaceHolder: String = "") : BaseKey() {
    override fun instantiateFragment(): Fragment = TasksFragment()

    override fun bindServices(serviceBinder: ServiceBinder) {
        with(serviceBinder) {
            add(TasksViewModel(
                Injector.get().tasksDataSource(),
                Injector.get().resources(),
                backstack,
                Injector.get().messageQueue(),
                getKey()
            ))
        }
    }

    override fun navigationViewId(): Int {
        return R.id.list_navigation_menu_item
    }

    override fun shouldShowUp(): Boolean {
        return false
    }

    override fun setupFab(fragment: Fragment, fab: FloatingActionButton) {
        fab.setImageResource(R.drawable.ic_add)
        fab.setOnClickListener { v: View? -> (fragment as TasksFragment).onAddNewTaskClicked() }
    }

    override val isFabVisible: Boolean
        get() = true
}